File: src/query.js
const fetch = require('node-fetch')
const https = require('https') 

const { google } = require('googleapis')
const { GoogleSpreadsheet } = require('google-spreadsheet')
const Twitter = require('twitter')

const jsonQuery = require('json-query') 
var dateFormat = require('dateformat')
     
const httpsOptions = {
    agent: new https.Agent({
      rejectUnauthorized: false
    })
};

(async () => {

    await fetch('https://data.ontario.ca/en/api/3/action/datastore_search?resource_id=455fd63b-603d-4608-8216-7d8647f43350&fields=Outcome1&limit=10000000', httpsOptions)
    // await fetch('https://data.ontario.ca/en/api/3/action/datastore_search?resource_id=455fd63b-603d-4608-8216-7d8647f43350&fields=Outcome1&limit=10', httpsOptions)
    // await fetch('https://data.ontario.ca/api/3/action/datastore_search?resource_id=ed270bb8-340b-41f9-a7c6-e8ef587e6d11&limit=1000000', httpsOptions)
    .then(response => response.json())
    .then(data => {

        var resolved = jsonQuery('result.records[*Outcome1=Resolved]', {
            data: data
        })

        var fatal = jsonQuery('result.records[*Outcome1=Fatal]', {
            data: data
        })

        console.log("Total Cases: " + data.result.total.toString());
        console.log("Resolved Cases: " + resolved.value.length.toString())
        console.log("Fatal Cases: " + fatal.value.length.toString())
        console.log("Active Cases: " + (data.result.total - resolved.value.length - fatal.value.length).toString())

        AccessSpreadsheet(data.result.total, resolved.value.length, fatal.value.length)

    })
    .catch(err => {
        console.log(err)
    })

})()

async function AccessSpreadsheet(todayTotalCases, todayTotalResolved, todayTotalFatal) {

    const creds = require('./client_secret.json');
    const doc = new GoogleSpreadsheet(creds.spreadsheet_url);

    await doc.useServiceAccountAuth({
        client_email: creds.client_email,
        private_key: creds.private_key,
    });

    await doc.loadInfo()

    const rows = await doc.sheetsByIndex[0].getRows()

    var today = dateFormat(new Date(), "mmmm dd yyyy")
    var yesterday = rows[rows.length - 1]["Date"]
    var yesterdayTotalCases = rows[rows.length - 1]["Total Cases"]

    if (yesterday != today && yesterdayTotalCases != todayTotalCases && todayTotalCases >= yesterdayTotalCases) {

        var activeCases = todayTotalCases - todayTotalResolved - todayTotalFatal
        var newCases = todayTotalCases - yesterdayTotalCases
        var newRecoveries = todayTotalResolved - rows[rows.length - 1]["Resolved Cases"]
        var newDeaths = todayTotalFatal - rows[rows.length - 1]["Deceased Cases"]
        var activeCaseDifference = activeCases - rows[rows.length - 1]["Active Cases"]

        doc.sheetsByIndex[0].addRow({

            "Date": today,
            "Total Cases": todayTotalCases,
            "New Cases": newCases,
            // 7 Day Avg
            "Resolved Cases": todayTotalResolved,
            "New Recoveries": newRecoveries,
            "Deceased Cases": todayTotalFatal,
            "New Deaths": newDeaths,
            "Active Cases": activeCases,
            "Active Case Difference": activeCaseDifference

        })

        doc.sheetsByIndex[0].saveUpdatedCells()

        const config = require("./TwitterAPIKeys")
        const twitterClient = new Twitter(config)
        twitterClient.post('statuses/update', {status:

            "Ontario COVID-19 data for today (" + dateFormat(new Date(), "mmmm dd yyyy") + "): \n" +
            newCases + " new cases \n" +
            newRecoveries + " recoveries \n" +
            newDeaths + " deaths \n" +
            activeCases + " active cases (" + (activeCaseDifference >= 0 ? "+" : "") + activeCaseDifference + ") \n\n" +

            "Visit https://covontario.ca to view the graph \n\n" +

            "#COVID19Ontario #COVID19 #COVID #Ontario"

            }, function(error, tweet, response){
            if(!error){
                console.log(error);
            }
        })

    }
    else if (rows[rows.length - 1]["Date"] === today) {

        var latestRow = rows[rows.length - 1]
        var firstRow = rows[0]

        if (!latestRow['New Tests'] ||
            !latestRow['Hospitalizations'] ||
            !latestRow['ICU'] ||
            !latestRow['ICU_Ventilated']) {
        
            await fetch('https://data.ontario.ca/api/3/action/datastore_search?resource_id=ed270bb8-340b-41f9-a7c6-e8ef587e6d11&limit=1&sort=%22Reported%20Date%22%20desc', httpsOptions)
            .then(response => response.json())
            .then(data => { 

                var latestData = data.result.records[data.result.records.length - 1]
                var newTests = latestData['Total tests completed in the last day']
                var totalTests = latestData['Total patients approved for testing as of Reporting Date']
                var hospitalizations = latestData['Number of patients hospitalized with COVID-19']
                var icu = latestData['Number of patients in ICU due to COVID-19']
                var icuVented = latestData['Number of patients in ICU on a ventilator due to COVID-19']

                var dateFromAPI = new Date(data.result.records[0]['Reported Date'])
                var today = new Date()

                var dateStringAPI = dateFromAPI.getDate() + '/' + dateFromAPI.getMonth() + '/' + dateFromAPI.getFullYear()
                var dateStringToday = today.getDate() + '/' + today.getMonth() + '/' + today.getFullYear()

                if (dateStringAPI === dateStringToday) {

                    latestRow['New Tests'] = newTests
                    firstRow['Total Tests'] = totalTests

                    latestRow['Hospitalizations'] = hospitalizations
                    latestRow['ICU'] = icu
                    latestRow['ICU_Ventilated'] = icuVented

                    latestRow.save()
                    firstRow.save()

                }
            })
        }

        if (!latestRow['Cases_Vax'] ||
            !latestRow['Cases_Unvax']  ||
            !latestRow['Cases_Partial'] ||
            !latestRow['Cases_Unknown']) {

            await fetch('https://data.ontario.ca/en/api/3/action/datastore_search?resource_id=eed63cf2-83dd-4598-b337-b288c0a89a16&limit=1&sort=Date%20desc', httpsOptions)
            .then(response => response.json())
            .then(data => {

                var vax = data.result.records[0]['covid19_cases_full_vac']
                var unvax = data.result.records[0]['covid19_cases_unvac']
                var partial = data.result.records[0]['covid19_cases_partial_vac']
                var unknown = data.result.records[0]['covid19_cases_vac_unknown']

                var dateFromAPI = new Date(data.result.records[0]['Date'])
                var today = new Date()

                var stringAPI = dateFromAPI.getDate() + '/' + dateFromAPI.getMonth() + '/' + dateFromAPI.getFullYear()
                var stringToday = today.getDate() + '/' + today.getMonth() + '/' + today.getFullYear()

                if (stringAPI === stringToday) {

                    latestRow['Cases_Vax'] = vax
                    latestRow['Cases_Unvax'] = unvax
                    latestRow['Cases_Partial'] = partial
                    latestRow['Cases_Unknown'] = unknown

                    latestRow.save()

                }
            })
        }
    }

    // Date
    // Total Cases
    // New Cases
    // 7 Day Average
    // Resolved Cases
    // New Recoveries
    // Deceased Cases
    // New Deaths
    // Active Cases
    // Active Case Difference
    // New Tests
    // % Positive
    // Total Tests

}


    

File: script/ruby_query/query.rb
#!/usr/bin/env ruby

require 'net/http'
require 'json'
require 'uri'
require 'bundler'
require 'date'

require './CaseData.rb'
Bundler.require

def GetData
    # url = 'https://data.ontario.ca/en/api/3/action/datastore_search?resource_id=455fd63b-603d-4608-8216-7d8647f43350&fields=Outcome1&limit=100'
    # url = 'https://data.ontario.ca/en/api/3/action/datastore_search?resource_id=455fd63b-603d-4608-8216-7d8647f43350&fields=Outcome1&limit=10000000'
    url = 'https://data.ontario.ca/en/api/3/action/datastore_search?resource_id=ed270bb8-340b-41f9-a7c6-e8ef587e6d11&limit=14&sort=Reported%20Date%20desc'
    uri = URI(url)
    response = Net::HTTP.get(uri)

    json = JSON.parse(response)
    records = json['result']['records']

    return records

    puts 'done'
end


def AccessSpreadsheet(data)

    dataQueue = Queue.new

    # Create google sheet context
    session = GoogleDrive::Session.from_service_account_key("../client_secret.json")
    spreadsheet = session.spreadsheet_by_title("Ontario COVID-19")
    worksheet = spreadsheet.worksheets.first
    rows = worksheet.rows
    totalRows = worksheet.rows.length

    # Create hash that maps each column name to an integer
    # Used because this sheet API doesnt allow accessing rows by key name, only by index
    count = 0
    headerRow = {}
    rows[0].each do |e|
        headerRow[e] = count
        count = count + 1
    end

    # puts headerRow

    data.reverse!
    beginInsert = false

    data.each do |d|

        year = d['Reported Date'][0,4].to_i
        month = d['Reported Date'][5,6].to_i
        day = d['Reported Date'][8,9].to_i

        parsed = Date.new(year, month, day)
        date = parsed.strftime('%B %d %Y')

        if rows[totalRows - 1][headerRow['Date']] == date
            beginInsert = true
            next
        elsif beginInsert == true
            obj = CreateCaseDataObject(d, date)
            WriteObjectToSpreadsheet(obj, worksheet, headerRow)
            PrintCaseDataObject(obj)
            dataQueue.enq(obj)
        end

    end
    
    worksheet.save
    return dataQueue

end

def WriteObjectToSpreadsheet(data, sheet, headerRow)

    rows = sheet.rows
    totalRows = sheet.rows.length

    newCases = data.total - rows[totalRows - 1][headerRow['Total Cases']].to_i
    newRecoveries = data.resolved - rows[totalRows - 1][headerRow['Resolved Cases']].to_i
    newFatalities = data.fatal - rows[totalRows - 1][headerRow['Deceased Cases']].to_i
    activeCaseDifference = data.active - rows[totalRows - 1][headerRow["Active Cases"]].to_i

    data.newCases = newCases
    data.newRecoveries = newRecoveries
    data.newFatalities = newFatalities
    data.activeCaseDifference = activeCaseDifference

    # Date
    # Total Cases  
    # New Cases
    # 7 Day Average	
    # Resolved Cases	
    # New Recoveries	
    # Deceased Cases	
    # New Deaths	
    # Active Cases	
    # Active Case Difference	
    # New Tests
    # % Positive	
    # Total Tests		
    # *** SPARE ***
    # Cases_NotFullyVaccinated	
    # Cases_Vax	
    # Cases_Boosted	
    # Cases_Unknown	
    # Hospitalizations	
    # ICU	ICU_Ventilated	
    # Cases_Unvax
    # Cases_Partial										

    sheet.insert_rows(totalRows + 1, [[
        data.date,
        data.total,
        data.newCases,
        '',
        data.resolved,
        data.newRecoveries,
        data.fatal,
        data.newFatalities,
        data.active,
        data.activeCaseDifference,
        data.newTests,
        '',
        data.totalTests,
        '', # SPARE
        '',
        '',
        '',
        '',
        data.hospital,
        data.icu,
        data.icuVented
    ]])
end

def CreateCaseDataObject(data, date)

    date = date
    total = data['Total Cases']
    resolved = data['Resolved']
    fatal = data['Deaths_New_Methodology']
    active = data['Confirmed Positive']
    newTests = data['Total tests completed in the last day']
    totalTests = data['Total patients approved for testing as of Reporting Date']
    hospitalizations = data['Number of patients hospitalized with COVID-19']
    icu = data['Number of patients in ICU due to COVID-19']
    icuVented = data['Number of patients in ICU on a ventilator due to COVID-19']

    caseData = CaseData.new(date, total, resolved, fatal, active, newTests, totalTests, hospitalizations, icu, icuVented)
    return caseData

end

def PrintCaseDataObject(data)

    puts data.inspect

end

def TweetDataQueue(client, dataQueue)

    while dataQueue.length > 0
        data = dataQueue.deq()
        Tweet(client, data)
    end

end

def Tweet(client, data)

    client.update(
        "Ontario COVID-19 case data for " + data.date + ": \n" +
        "Cases: " + data.newCases.to_s + "\n" +
        "Deaths: " + data.newFatalities.to_s + "\n" +
        "Recoveries: " + data.newRecoveries.to_s + "\n" +
        "Active Cases: " + data.active.to_s + " (" + (data.activeCaseDifference >= 0 ? '+' : '') + data.activeCaseDifference.to_s + ") \n\n" +

        "Visit https://covontario.ca to view additional data \n\n" +

        "#COVID19Ontario #COVID19 #COVID #Ontario #Covontario"
    )

end

def GetTwitterClient

    creds = File.read('../TwitterAPIKeys_Ruby.json')
    credHash = JSON.parse(creds)
    
    client = Twitter::REST::Client.new do |config|
        config.consumer_key        = credHash['consumer_key']
        config.consumer_secret     = credHash['consumer_secret']
        config.access_token        = credHash['access_token_key']
        config.access_token_secret = credHash['access_token_secret']
    end

    return client

end

data = GetData()
client = GetTwitterClient()
dataQueue = AccessSpreadsheet(data)
TweetDataQueue(client, dataQueue)
